PWGMM/Tasks/vertexing-fwd.cxx

```diff
@@ -0,0 +1,188 @@
+     //{"TracksDCAXYOther", "; DCA_{xy}^{best, false} (cm); counts", {HistType::kTH1F, {{100, -1, 10}}}},
+     {"TracksDCAXYBestFalse", "; DCA_{xy}^{best, false} (cm); counts", {HistType::kTH1F, {{100, -1, 10}}}},
+     {"EfficiencyZvtx", "; z vertex; Efficiency", {HistType::kTProfile, {{100, -30, 30}}}},
+     {"DeltaZvtx", "; #delta z (cm); counts", {HistType::kTH1F, {{400, -20, 20}}}},
+     {"DeltaZvtxBest", "; #delta z = z_{best} - z_{true} (cm); counts", {HistType::kTH1F, {{400, -20, 20}}}},
+     //{"DeltaDCAminNcoll", "; N_{coll}; #delta DCA = DCA_{xy}^{min} - DCA_{xy}^{other} (cm); counts", {HistType::kTH2F, {{10, -0.5, 9.5}, {100, -0.5, 9.5}}}},
+     {"CorrectMatch", "; Matching value; counts", {HistType::kTH1F, {{4, -0.5, 3.5}}}}}};
+
+  void init(InitContext&)
+  {
+    auto hstat = registry.get<TH1>(HIST("CorrectMatch"));
+    auto* x = hstat->GetXaxis();
+    x->SetBinLabel(1, "Incorrect match");
+    x->SetBinLabel(2, "Correct match");
+    x->SetBinLabel(3, "Total w N_{coll} > 0");
+    x->SetBinLabel(4, "Total w N_{coll} #geq 0");
+  }
+
+  int getIndexBestCollision(std::vector<double> vecOfDCA, int method = 0)
+  {
+    int indice = 0;
+    if (method == 0) {
+      indice = std::distance(vecOfDCA.begin(), std::min_element(vecOfDCA.begin(), vecOfDCA.end()));
+    }
+    return indice;
+  }
+
+  void process(aod::AmbiguousMFTTracks const& ambitracks, aod::BCs const& bcs, soa::Join<o2::aod::MFTTracks, aod::McMFTTrackLabels> const& tracks, soa::Join<aod::Collisions, aod::McCollisionLabels> const& collisions, aod::McParticles const& mcParticles, aod::McCollisions const& mcCollisions)
+  {
+    for (auto& ambitrack : ambitracks) {
+      vecCollForAmb.clear();
+      vecDCACollForAmb.clear();
+      vecAmbTrack.clear();
+      vecZposCollForAmb.clear();
+
+      double value = 0.0;    // matching value for collision association to an ambiguous track
+      double zVtxMCAmbi = 0; // z vertex associated to the mc collision
+      int mcCollAmbiID = -1; // mc value for the collision containing the ambiguous track
+
+      auto track = ambitrack.mfttrack_as<MFTTracksLabeled>(); // Obtain the MFT ambiguous track with the MC labels
+      // auto extAmbiTrackid = ambitrack.mfttrackId(); // Global index of the MFT ambiguous track
+
+      if (!track.has_mcParticle()) {
+        LOGF(warning, "No MC particle for ambiguous track, skip...");
+        continue;
+      }
+      auto particle = track.mcParticle();
+      mcCollAmbiID = particle.mcCollisionId();
+      zVtxMCAmbi = particle.mcCollision().posZ();
+
+      // Fill the std::vector for ambiguous tracks with the quantities needed
+      vecAmbTrack.push_back(track.x());
+      vecAmbTrack.push_back(track.y());
+      vecAmbTrack.push_back(track.phi());
+      vecAmbTrack.push_back(track.tgl());
+      vecAmbTrack.push_back(track.signed1Pt());
+      vecAmbTrack.push_back(track.z());
+      vecAmbTrack.push_back(track.chi2());
+
+      for (auto& collision : collisions) {
```

Comment:
Looping over the whole collision table is a bit heavy, there is, however `Run3MatchedToBCSparse` table that is compatible with BCs table and can be joined to it, that also has an index to collisions. If you use ambiguous track's index that refers to it, you can directly retrieve a slice of the joined BC+Run3MatchedToBCSparse table and go directly to collisions (if any) from it.

Reply:
`bcIds().size()` returns the size of the _underlying index array_ which is always 2 - it contains two integers, the start and the end of a slice. The actual getter `bc_as<>` returns the subtable, size of it is what you need here.

Reply:
Thanks for your answer Anton ! We then have a bigger problem, the code crashes when we try to get any information from collision1, such as collision1.posZ() or such ... Even though we checked that the collisionId needed to be >0 (-1 means no collision associated to this BC)
I really don't see why
Any idea of a security we could implement to avoid this issue ?

Reply:
Which version of code is crashing? There can be several reasons for the crash when dereferencing the index.

Reply:
Hi Anton, Robin just pushed the version of the code which is crashing, it crashes when we try to access collision.posZ() (just after the LOG 1bis)
Any idea of how to fix this ? The error doesn't seem really helpful
<img width="1422" alt="Screenshot 2022-05-09 at 09 54 28" src="https://user-images.githubusercontent.com/83958698/167366251-1892fcc0-8bf4-4032-b6fe-5829733dca35.png">

Reply:
I see now, thank you. You need to check if there is a collision for this bc, using `bcambi.has_collision()`.

Reply:
Ok thanks a lot ! Will do